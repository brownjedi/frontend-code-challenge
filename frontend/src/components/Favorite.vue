<template>
  <div :class="styles['favorite-container']">
    <div
      :class="[
        isFavorite ? styles.favorite : styles['not-favorite'],
        large ? styles.large : styles.small,
        animated && styles.gelatine,
      ]"
      @click.stop="handleClick"
      @animationend="animated = false"
    ></div>
  </div>
</template>

<script lang="ts">
import { defineComponent, ref, useCssModule } from 'vue'

export default defineComponent({
  props: {
    isFavorite: {
      type: Boolean,
      required: true,
    },
    large: {
      type: Boolean,
      default: false,
    },
  },
  emits: ['click'],
  setup(_, { emit }) {
    const styles = useCssModule()
    const animated = ref(false)

    function handleClick() {
      animated.value = true
      emit('click')
    }

    return { styles, handleClick, animated }
  },
})
</script>

<style lang="scss" module>
.favorite-container {
  align-self: flex-end;
  display: flex;
  align-items: flex-end;
  cursor: pointer;

  & > * {
    flex: 1;
  }

  .small {
    width: 1.75rem;
    height: 1.75rem;
  }

  .large {
    width: 2.75rem;
    height: 2.75rem;
  }

  .favorite {
    background: url('/img/pokeball_color.svg') no-repeat;
  }

  .not-favorite {
    background: url('/img/pokeball_bw.svg') no-repeat;
  }

  .gelatine {
    animation: gelatine 0.5s;
  }

  @keyframes gelatine {
    from,
    to {
      transform: scale(1, 1);
    }
    25% {
      transform: scale(0.9, 1.1);
    }
    50% {
      transform: scale(1.1, 0.9);
    }
    75% {
      transform: scale(0.95, 1.05);
    }
  }
}
</style>
