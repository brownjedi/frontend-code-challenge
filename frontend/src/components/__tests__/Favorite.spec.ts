import { shallowMount } from '@vue/test-utils'
import Favorite from '@/components/Favorite.vue'

describe('components/Favorite.vue', () => {
  it('has the expected html structure', () => {
    const wrapper = shallowMount(Favorite, {
      props: {
        isFavorite: true,
      },
    })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('has correct class when setting isFavorite', async () => {
    const wrapper = shallowMount(Favorite, {
      props: {
        isFavorite: true,
      },
    })
    expect(wrapper.find('.favorite').exists()).toBe(true)
    expect(wrapper.find('.not-favorite').exists()).toBe(false)

    await wrapper.setProps({ isFavorite: false })
    expect(wrapper.find('.favorite').exists()).toBe(false)
    expect(wrapper.find('.not-favorite').exists()).toBe(true)
  })

  it('has correct class when setting large', async () => {
    const wrapper = shallowMount(Favorite, {
      props: {
        isFavorite: true,
        large: false,
      },
    })

    expect(wrapper.find('.large').exists()).toBe(false)
    expect(wrapper.find('.small').exists()).toBe(true)

    await wrapper.setProps({ isFavorite: true, large: true })

    expect(wrapper.find('.large').exists()).toBe(true)
    expect(wrapper.find('.small').exists()).toBe(false)
  })

  it('emits when clicked', async () => {
    const wrapper = shallowMount(Favorite, {
      props: {
        isFavorite: true,
      },
    })

    expect(wrapper.emitted().click).toBeUndefined()

    await wrapper.find('.favorite').trigger('click')

    expect(wrapper.emitted().click.length).toEqual(1)
  })
})
