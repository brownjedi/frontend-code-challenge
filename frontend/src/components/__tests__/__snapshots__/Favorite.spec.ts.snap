// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Favorite.vue has the expected html structure 1`] = `
<div
  class="favorite-container"
>
  <div
    class="favorite small"
  />
</div>
`;
